assembly/src/assembler/mast_forest_builder.rs

```diff
@@ -20,28 +29,112 @@ impl MastForestBuilder {
+    /// Returns a reference to the procedure with the specified MAST root, or None
+    /// if such a procedure is not present in this MAST forest builder.
     #[inline(always)]
-    pub fn find_procedure_root(&self, digest: RpoDigest) -> Option<MastNodeId> {
+    pub fn find_procedure(&self, mast_root: &RpoDigest) -> Option<Arc<Procedure>> {
+        self.proc_gid_by_hash.get(mast_root).and_then(|gid| self.get_procedure(*gid))
+    }
+
+    /// Returns the [`MastNodeId`] of the procedure associated with a given MAST root, or None
+    /// if such a procedure is not present in this MAST forest builder.
+    #[inline(always)]
+    pub fn find_procedure_node_id(&self, digest: RpoDigest) -> Option<MastNodeId> {
         self.mast_forest.find_procedure_root(digest)
     }
+
+    /// Returns the [`MastNode`] for the provided MAST node ID, or None if a node with this ID is
+    /// not present in this MAST forest builder.
+    pub fn get_mast_node(&self, id: MastNodeId) -> Option<&MastNode> {
+        self.mast_forest.get_node_by_id(id)
+    }
+}
+
+impl MastForestBuilder {
+    /// Inserts a procedure into this MAST forest builder.
+    ///
+    /// If the procedure with the same ID already exists in this forest builder, this will have
+    /// no effect.
+    pub fn insert_procedure(
+        &mut self,
+        gid: GlobalProcedureIndex,
+        procedure: Procedure,
+    ) -> Result<(), AssemblyError> {
+        let proc_root = self.mast_forest[procedure.body_node_id()].digest();
+
+        // Check if an entry is already in this cache slot.
+        //
+        // If there is already a cache entry, but it conflicts with what we're trying to cache,
+        // then raise an error.
+        if let Some(cached) = self.procedures.get(&gid) {
+            let cached_root = self.mast_forest[cached.body_node_id()].digest();
+            if cached_root != proc_root || cached.num_locals() != procedure.num_locals() {
+                return Err(AssemblyError::ConflictingDefinitions {
+                    first: cached.fully_qualified_name().clone(),
+                    second: procedure.fully_qualified_name().clone(),
+                });
+            }
+
+            // The global procedure index and the MAST root resolve to an already cached version of
+            // this procedure, nothing to do.
+            //
+            // TODO: We should emit a warning for this, because while it is not an error per se, it
+            // does reflect that we're doing work we don't need to be doing. However, emitting a
+            // warning only makes sense if this is controllable by the user, and it isn't yet
+            // clear whether this edge case will ever happen in practice anyway.
+            return Ok(());
+        }
+
+        // We don't have a cache entry yet, but we do want to make sure we don't have a conflicting
+        // cache entry with the same MAST root:
+        if let Some(cached) = self.find_procedure(&proc_root) {
```

Comment:
I think we need to discuss adding a new node type to MAST that represents procedure entry, specifically to ensure that the number of procedure locals is represented in the digest. This also lets us recover essential procedure metadata when loading a library from MAST (such as number of procedure locals).